README.md
# flutter_treeview

A hierarchical data widget for your flutter apps. 

It offers a number of options for customizing the appearance and handling user interaction.

It also offers some convenience methods for importing data into the tree.


## Features

* Separately customize child and parent labels
* Add any icon to a node
* Choose from four different expander icons and several modifiers for adjusting shape, outline, and fill. 
* Import data from a Map 
* Includes ability to handle expandChange, tap, and double tap user interactions
* Includes convenience methods for adding, updating and deleting nodes


## Sample Code
### Creating a TreeView
```dart
List<Node> nodes = [
 Node(
   label: 'Documents',
   key: 'docs',
   expanded: true,
   icon: NodeIcon(
     codePoint:
         docsOpen ? Icons.folder_open.codePoint : Icons.folder.codePoint,
     color: "blue",
   ),
   children: [
     Node(
         label: 'Job Search',
         key: 'd3',
         icon: NodeIcon.fromIconData(Icons.input),
         children: [
           Node(
               label: 'Resume.docx',
               key: 'pd1',
               icon: NodeIcon.fromIconData(Icons.insert_drive_file)),
           Node(
               label: 'Cover Letter.docx',
               key: 'pd2',
               icon: NodeIcon.fromIconData(Icons.insert_drive_file)),
         ]),
     Node(
       label: 'Inspection.docx',
       key: 'd1',
     ),
     Node(
         label: 'Invoice.docx',
         key: 'd2',
         icon: NodeIcon.fromIconData(Icons.insert_drive_file)),
   ],
 ),
 Node(
     label: 'MeetingReport.xls',
     key: 'mrxls',
     icon: NodeIcon.fromIconData(Icons.insert_drive_file)),
 Node(
     label: 'MeetingReport.pdf',
     key: 'mrpdf',
     icon: NodeIcon.fromIconData(Icons.insert_drive_file)),
 Node(
     label: 'Demo.zip',
     key: 'demo',
     icon: NodeIcon.fromIconData(Icons.archive)),
];
TreeViewController _treeViewController = TreeViewController(children: nodes);
TreeView(
  controller: _treeViewController,
  allowParentSelect: false,
  supportParentDoubleTap: false,
  onExpansionChanged: _expandNodeHandler,
  onNodeTap: (key) {
    setState(() {
      _treeViewController = _treeViewController.copyWith(selectedKey: key);
    });
  },
  theme: treeViewTheme
),
```
_The TreeView requires that the onExpansionChange property updates the expanded
node so that the tree is rendered properly.

### Creating a theme
```dart
TreeViewTheme _treeViewTheme = TreeViewTheme(
  expanderTheme: ExpanderThemeData(
    type: ExpanderType.caret,
    modifier: ExpanderModifier.none,
    position: ExpanderPosition.start,
    color: Colors.red.shade800,
    size: 20,
  ),
  labelStyle: TextStyle(
    fontSize: 16,
    letterSpacing: 0.3,
  ),
  parentLabelStyle: TextStyle(
    fontSize: 16,
    letterSpacing: 0.1,
    fontWeight: FontWeight.w800,
    color: Colors.red.shade600,
  ),
  iconTheme: IconThemeData(
    size: 18,
    color: Colors.grey.shade800,
  ),
  colorScheme: ColorScheme.light(),
)
```

### Using custom data
The Node class supports the use of custom data. You can use the data property on the Node instance to store data that you want to easily retrieve.
```dart
class Person {
  final String name;
  final List<Animal> pets;

  Person({this.name, this.pets});
}

class Animal {
  final String name;

  Animal({this.name});
}

Animal otis = Animal(name: 'Otis');
Animal zorro = Animal(name: 'Zorro');
Person lukas = Person(name: 'Lukas', pets: [otis, zorro]);

List<Node> nodes = [
  Node<Person>(
    label: 'Lukas',
    key: 'lukas',
    data: lukas,
    children: [
      Node<Animal>(
        label: 'Otis',
        key: 'otis',
        data: otis,
      ),      
      //<T> is optional but recommended
      Node(
        label: 'Zorro',
        key: 'zorro',
        data: zorro,
      ),
    ]
  ),
];
TreeViewController _treeViewController = TreeViewController(children: nodes);
TreeView(
  controller: _treeViewController,
  onNodeTap: (key) {
    Node selectedNode = _treeViewController.getNode(key);
    Person selectedModel = selectedNode.data;
  },
),
```


## Getting Started

For help getting started with this widget, view our 
[online documentation](https://bitbucket.org/kevinandre/flutter_treeview/wiki/Home) or view the
full [API reference](https://pub.dev/documentation/flutter_treeview/latest/).

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md
## [0.6.0+1]

### Added
* Added support for importing data property during JSON and Map load

## [0.5.0+1]

### Added
* Added support for using shrinkWrap, primary, and physics property on TreeView 

## [0.4.2+1]

### Added
* Added support for using external font packages

## [0.4.1+1]

### Updated
* Updated TreeView widget so that it inherits the ThemeData from context

## [0.4.0+1]

### Added
* Added expandToNode method to TreeViewController to support expanding all nodes down to specified node. Returns List<Node>.
* Added collapseToNode method to TreeViewController to support collapsing all nodes down to specified node. Returns List<Node>.
* Added withExpandToNode method to TreeViewController to support expanding all nodes down to specified node. Returns TreeViewController.
* Added withCollapseToNode method to TreeViewController to support expanding all nodes down to specified node. Returns TreeViewController.

## [0.3.0+1]

### Added
* Added generic data property to Node class to support the use of custom data

## [0.2.0+1]

### Updated
* Added animation controller dispose to TreeNode to prevent memory leaks

### Added
* Added new dense property to TreeViewTheme
* Added new loadJSON and loadMap convenience methods to TreeViewController for data loading
* Added new convenience methods to TreeViewController: toggleNode, withToggleNode, selectedNode

## [0.1.0+2]

### Updated
* Updated links to repository documentation
* Cleaned up warnings

## [0.0.4+1]

### Updated
* Added logic to update TreeNode when expanded programmatically
* Fixed issue with adding new node to a TreeNode with new children

## [0.0.3+7]

### Added
* Added api documentation

### Updated
* Added parentLabelStyle to TreeViewTheme to support separate styling for parent node

## [0.0.2+1]

### Added
* Added ExpanderModifier

### Updated
* Updated open source license
* Simplified ExpanderType
* Refactored TreeNodeExpander class and added animation to icon
* Updated default expander size

### Removed
* Removed custom TreeView font

## [0.0.1]

* Initial package release


example/lib/main.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_treeview/tree_view.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'TreeView Example',
      home: MyHomePage(title: 'TreeView Example'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  String _selectedNode;
  List<Node> _nodes;
  TreeViewController _treeViewController;
  bool docsOpen = true;
  final Map<ExpanderPosition, Widget> expansionPositionOptions = const {
    ExpanderPosition.start: Text('Start'),
    ExpanderPosition.end: Text('End'),
  };
  final List<Map<String, dynamic>> _expanderStyleOptions = [
    {"value": ExpanderStyle.arrow, "label": "Arrow"},
    {"value": ExpanderStyle.box, "label": "Box"},
    {"value": ExpanderStyle.chevron, "label": "Chevron"},
    {"value": ExpanderStyle.circle, "label": "Circle"},
    {"value": ExpanderStyle.filledCircle, "label": "Filled Circle"},
    {"value": ExpanderStyle.longArrow, "label": "Long Arrow"},
  ];
  ExpanderPosition _expanderPosition = ExpanderPosition.start;
  bool _allowParentSelect = false;
  bool _supportParentDoubleTap = false;
  int _expanderStyleIndex = 0;

  @override
  void initState() {
    _nodes = [
      Node(
        label: 'documents',
        key: 'docs',
        expanded: docsOpen,
        icon: NodeIcon(
          codePoint:
              docsOpen ? Icons.folder_open.codePoint : Icons.folder.codePoint,
          color: "blue",
        ),
        children: [
          Node(
              label: 'personal',
              key: 'd3',
              icon: NodeIcon.fromIconData(Icons.input),
              children: [
                Node(
                    label: 'Resume.docx',
                    key: 'pd1',
                    icon: NodeIcon.fromIconData(Icons.insert_drive_file)),
                Node(
                    label: 'Cover Letter.docx',
                    key: 'pd2',
                    icon: NodeIcon.fromIconData(Icons.insert_drive_file)),
              ]),
          Node(
            label: 'Inspection.docx',
            key: 'd1',
//          icon: NodeIcon.fromIconData(Icons.insert_drive_file),
          ),
          Node(
              label: 'Invoice.docx',
              key: 'd2',
              icon: NodeIcon.fromIconData(Icons.insert_drive_file)),
        ],
      ),
      Node(
          label: 'MeetingReport.xls',
          key: 'mrxls',
          icon: NodeIcon.fromIconData(Icons.insert_drive_file)),
      Node(
          label: 'MeetingReport.pdf',
          key: 'mrpdf',
          icon: NodeIcon.fromIconData(Icons.insert_drive_file)),
      Node(
          label: 'Demo.zip',
          key: 'demo',
          icon: NodeIcon.fromIconData(Icons.archive)),
    ];
    _treeViewController = TreeViewController(
      children: _nodes,
      selectedKey: _selectedNode,
    );
    super.initState();
  }

  ListTile _makeExpanderPosition() {
    return ListTile(
      title: Text('Expander Position'),
      trailing: CupertinoSlidingSegmentedControl(
        children: expansionPositionOptions,
        groupValue: _expanderPosition,
        onValueChanged: (ExpanderPosition newValue) {
          setState(() {
            _expanderPosition = newValue;
          });
        },
      ),
    );
  }

  SwitchListTile _makeAllowParentSelect() {
    return SwitchListTile.adaptive(
      title: Text('Allow Parent Select'),
      value: _allowParentSelect,
      onChanged: (v) {
        setState(() {
          _allowParentSelect = v;
        });
      },
    );
  }

  SwitchListTile _makeSupportParentDoubleTap() {
    return SwitchListTile.adaptive(
      title: Text('Support Parent Double Tap'),
      value: _supportParentDoubleTap,
      onChanged: (v) {
        setState(() {
          _supportParentDoubleTap = v;
        });
      },
    );
  }

  ListTile _makeExpanderType() {
    return ListTile(
      title: Text('Expander Style'),
      trailing: SizedBox(
        width: 150,
        child: Row(
          mainAxisSize: MainAxisSize.min,
          mainAxisAlignment: MainAxisAlignment.end,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: <Widget>[
            Text(
              _expanderStyleOptions.elementAt(_expanderStyleIndex)['label'],
              style: Theme.of(context).textTheme.caption,
            ),
            Icon(
              Icons.chevron_right,
              color: Theme.of(context).primaryColor,
            ),
          ],
        ),
      ),
      onTap: () async {
        final int styleIndex = await showCupertinoDialog(
            context: context,
            builder: (context) {
              Map<int, Map<String, dynamic>> _options =
                  _expanderStyleOptions.asMap();
              return SimpleDialog(
                titlePadding: EdgeInsets.only(left: 10),
                title: Row(
                  children: <Widget>[
                    Expanded(child: Text('Expander Style')),
                    IconButton(
                      icon: Icon(Icons.clear),
                      onPressed: () =>
                          Navigator.pop(context, _expanderStyleIndex),
                    ),
                  ],
                ),
                children: _options.keys.map((int itemIndex) {
                  Map<String, dynamic> item =
                      _expanderStyleOptions.elementAt(itemIndex);
                  return SimpleDialogOption(
                    child: Row(
                      mainAxisAlignment: MainAxisAlignment.spaceBetween,
                      children: <Widget>[
                        Icon(
                          Icons.check,
                          color: _expanderStyleIndex == itemIndex
                              ? Colors.blue
                              : Colors.transparent,
                        ),
                        Expanded(child: Text(item['label'])),
                      ],
                    ),
                    onPressed: () {
                      Navigator.pop(context, itemIndex);
                    },
                  );
                }).toList(),
              );
            });
        if (styleIndex != null)
          setState(() {
            _expanderStyleIndex = styleIndex;
          });
      },
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
        elevation: 0,
      ),
      body: Container(
        color: Colors.grey.shade200,
        padding: EdgeInsets.all(20),
        height: double.infinity,
        child: Column(
          children: <Widget>[
            Container(
              height: 300,
              child: Column(
                children: <Widget>[
                  _makeExpanderPosition(),
                  _makeAllowParentSelect(),
                  _makeSupportParentDoubleTap(),
                  _makeExpanderType(),
                ],
              ),
            ),
            Expanded(
              child: Container(
                decoration: BoxDecoration(
                  color: Colors.white,
                  borderRadius: BorderRadius.circular(10),
                ),
                padding: EdgeInsets.all(10),
                child: TreeView(
                  controller: _treeViewController,
                  allowParentSelect: _allowParentSelect,
                  supportParentDoubleTap: _supportParentDoubleTap,
                  onExpansionChanged: (key, expanded) =>
                      _expandNode(key, expanded),
                  onNodeSelect: (key) {
                    debugPrint('Selected: $key');
                    setState(() {
                      _treeViewController =
                          _treeViewController.copyWith(selectedKey: key);
                    });
                  },
                  theme: TreeViewTheme(
                    expanderStyle: _expanderStyleOptions
                        .elementAt(_expanderStyleIndex)['value'],
                    position: _expanderPosition,
                    labelStyle: TextStyle(
//              fontFamily: 'Times',
                      fontSize: 16,
                      letterSpacing: 0.3,
                    ),
                    iconTheme: IconThemeData(
                      size: 18,
                      color: Colors.grey.shade800,
                    ),
                    colorScheme:
                        Theme.of(context).brightness == Brightness.light
                            ? ColorScheme.light(
                                primary: Colors.blue.shade50,
                                onPrimary: Colors.grey.shade900,
                                background: Colors.transparent,
                                onBackground: Colors.black,
                              )
                            : ColorScheme.dark(
                                primary: Colors.black26,
                                onPrimary: Colors.white,
                                background: Colors.transparent,
                                onBackground: Colors.white70,
                              ),
                  ),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }

  _expandNode(String key, bool expanded) {
    String msg = '${expanded ? "Expanded" : "Collapsed"}: $key';
    debugPrint(msg);
    Node node = _treeViewController.getNode(key);
    if (node != null) {
      List<Node> updated;
      if (key == 'docs') {
        updated = _treeViewController.updateNode(
          key,
          node.copyWith(
              expanded: expanded,
              icon: NodeIcon(
                codePoint: expanded
                    ? Icons.folder_open.codePoint
                    : Icons.folder.codePoint,
                color: expanded ? "blue600" : "grey700",
              )),
        );
      } else {
        updated = _treeViewController.updateNode(
            key, node.copyWith(expanded: expanded));
      }
      setState(() {
        if (key == 'docs') docsOpen = expanded;
        _treeViewController = _treeViewController.copyWith(children: updated);
      });
    }
  }
}
